### DISCUSSION TITLE
Confusion Matrix and $predicted from rfsrc

### DISCUSSION THREAD

**Discussion by RacioBR** [kogalur/randomForestSRC#231] [Q&A] [answered]
It is the first time that I am using Random Forest and pretty much my first machine learning model. I don't know if this is a dumb question or not. So, sorry in advance.

I want to use the $predicted information in a trained random forest using rfsrc in my application but when I compared the results of the probability predicted ($predicted) and the confusion matrix it doesn't make much sense.

See below:

Confusion matrix:
                                predicted
observed.            Cat.1.          Cat.2.   class.error

Cat.1                 59,674         86         0.0014

Cat.2                   3,578         78        0.9787

Table of Probabilities Predicted (frequencies)
                              Observed cases
Prob. Cat. 2.        Cat.1.             Cat.2
  0 - 10%                 57,681                0
10 - 20%                 2,007                  0
20 - 30%                    64                    0
30 - 40%                     7                     0
40 - 50%                     1                     1
50 - 60%                     0                    46
60 - 70%                     0                2,133
70 - 80%                     0                1,354
80 - 90%                     0                  119
90 - 100%                   0                    3

![20220422_062141.jpg](https://user-images.githubusercontent.com/31930229/164677356-54dff20a-9493-47aa-9f85-314d4b8ed253.jpg)

The table of Probabilities Predicted for Cat. 2 is pretty good but the confusion matrix is the opposite. I could not understand.

I am running a R script in Linux terminal. See some results below.

![20220421_210418.jpg](https://user-images.githubusercontent.com/31930229/164570085-d084058a-7950-49a6-91f3-29327f82c8a5.jpg)

**Comment by ishwaran** [comment, marked as answer]
The confusion matrix is constructed using the predicted class labels and they in turn are calculated from the predicted probabilities using Bayes rule (assignment to the class with highest probability).  I use the iris data below to show you how the confusion matrix is obtained from the out-of-bag class label predicted values.  Regarding Bayes rule, this is something you need to read up and understand.

In the following `o$yvar` is the response from the analysis (Species) and `o$class.oob` is the out-of-bag predicted class label:

```
> o=rfsrc(Species~.,iris)
> table(o$yvar, o$class.oob)
            
             setosa versicolor virginica
  setosa         49          1         0
  versicolor      0         47         3
  virginica       0          3        47

> o
                         Sample size: 150
           Frequency of class labels: 50, 50, 50
                     Number of trees: 500
           Forest terminal node size: 1
       Average no. of terminal nodes: 9.57
No. of variables tried at each split: 2
              Total no. of variables: 4
       Resampling used to grow trees: swor
    Resample size used to grow trees: 95
                            Analysis: RF-C
                              Family: class
                      Splitting rule: gini *random*
       Number of random split points: 10
                   (OOB) Brier score: 0.02392565
        (OOB) Normalized Brier score: 0.10766541
                           (OOB) AUC: 0.9928
   (OOB) Requested performance error: 0.04666667, 0.02, 0.06, 0.06

Confusion matrix:

            predicted
  observed   setosa versicolor virginica class.error
  setosa         49          1         0        0.02
  versicolor      0         47         3        0.06
  virginica       0          3        47        0.06

      (OOB) Misclassification rate: 0.04666667

```

**Reply by a-gialluisi** [reply]
Hi all, I am using $predicted within a rfsrc model.
Just to clarify, $predicted should be interpreted as % probability of event, right? 
But what is the timeframe of reference? Shall we consider the median follow-up time of all the subjects analyzed or is there any other way to define a time that I am not aware of?

Thanks,
Alessandro

**Reply by ishwaran** [reply]
It's not clear what you are asking.  This thread is about classification but the fact that you mention follow-up time seems to suggest you are interested in survival.  I suggest you look at the [getting started vignette on prediction](https://www.randomforestsrc.org/articles/getstarted.html#prediction-and-terminal-node-statistics-1) where the predicted value that is returned by a call to `rfsrc` is defined for each family.  See the table there for details.

**Reply by a-gialluisi** [reply]
HI, you are right, I am running a survival.
In the vigette it reads 
_"For [survival](https://www.randomforestsrc.org/articles/survival.html), the TNS is the Kaplan-Meier estimator and the Nelson-Aalen cumulative hazard function (CHF) at the time points of interest specified by the user, or as determined by the package if not specified [17]."_

But I still have a some doubts which I need to clarify:
- How can I determine whether it is a Kaplan Meier survival or a CHF?
- I did not specify the time points for the KMS/CHF estimate..how can I infer the default value? I had a look at ref 17 and was not able to find it.

I posted the answer here since I believe it is still in line with the title of the discussion, but please let me know whether you find it more appropriate to open a new discussion on the topic. Thanks a lot for your help.

AG

**Reply by ishwaran** [reply]
For survival analysis, `rfsrc` returns two objects: `$survival.oob` and `$chf.oob`, where the first is the ensemble survival function using KM tree estimators and the second is the ensemble CHF using Nelson-Aalen tree estimators.  

The default time values are stored in `$time.interest` which are the ordered unique death times.  The choice of which unique death times to be used is controlled by the option `time`.  From the help file:

```
ntime: Integer value used for survival to constrain ensemble
          calculations to an ‘ntime’ grid of time points over the
          observed event times.  Alternatively if a vector of values of
          length greater than one is supplied, it is assumed these are
          the time points to be used (these will be adjusted to match
          closest observed event times).  Setting ‘ntime’ to zero (or
          NULL) uses all observed event times.
```

**Reply by a-gialluisi** [reply]
Hi, I am afraid I am missing something.
So I checked the objects you specified ($survival.oob and $chf.oob) and both have the "NULL" value. What does that mean exactly?
Also, I rescued the _$time.interest_ values and computed a median  to have a follow-up time range for my estimation of risk. Is that correct?
To clarify, what I would like to do is making an estimate of probability of disease occurence over a given time range for the disease risk analyzed in my survival random forest. That is, for each subject I would like to state that (e.g.)
"subject A has a x% probability of incident disease in a y years time range".
What values shall I take as representative of x and y?

Thanks a lot for your help.
Rgds,
Alessandro
